pkg/deploy/controller/deployer/deployer_controller.go

```diff
@@ -135,22 +139,81 @@ func (c *DeploymentController) handle(deployment *v1.ReplicationController, will
 			}
 
 			// Generate a deployer pod spec.
-			deployerPod, err := c.makeDeployerPod(deployment)
+			deployerPod, err = c.makeDeployerPod(deployment)
 			if err != nil {
 				return fatalError(fmt.Sprintf("couldn't make deployer pod for %q: %v", deployutil.LabelForDeploymentV1(deployment), err))
 			}
+
+			// We need to transition the deployment to "pending" to make sure we can't ever create the deployer pod twice.
+			// If we would fail to update phase after and the deployer pod would have gotten deleted in the meantime
+			// we would have recreated it here.
+			// The disadvantage is that if the Create call fails we will never create the deployer pod.
```

Comment:
It's interesting that we have an old controller loop observing that it needs to create a pod.  It's a retry?  Or something else?  

That doesn't mean this might not be the right choice, but I'm concerned about retry failing.  The controller itself seems like it has a concurrency bug that should be fixed, rather than changing how this code runs.

Ie:

1. Observe need for new deployment.
2. Observe running rcs - after this point, the controller should never try to start an older rc than the one it sees
3. Cancel older deployment pods
4. Create deployment pod for most recent, but only if 2 still holds

Can we correct 4 to ensure the invariant from 2 inside the controller is repsect?  This bug implies 4 is proceeding even though we started back at 1

Reply:
@smarterclayton I don't think that can be done. Controller processes queue in parallel for different keys (RCs). The issue is that you can't do atomic operation on more that one object. 

See my comment in main section. (I didn't wanted it to get lost/hidden here in case of rebase.)